src/vmm/src/memory_dump.rs

```diff
@@ -3,52 +3,113 @@
+    pub regions: Vec<GuestMemoryRegionState>,
+}
+
+/// Defines the interface for snapshotting memory.
+pub trait SnapshotMemory
+where
+    Self: Sized,
+{
+    /// Describes GuestMemoryMmap through a GuestMemoryState struct.
+    fn describe(&self) -> GuestMemoryState;
+    /// Dumps all contents of GuestMemoryMmap to a writer.
     fn dump<T: std::io::Write>(&self, writer: &mut T) -> std::result::Result<(), Error>;
+    /// Dumps all pages of GuestMemoryMmap present in `dirty_bitmap` to a writer.
     fn dump_dirty<T: std::io::Write + std::io::Seek>(
         &self,
         writer: &mut T,
         dirty_bitmap: &DirtyBitmap,
     ) -> std::result::Result<(), Error>;
+    /// Creates a GuestMemoryMmap give a `file` containing the data
+    /// and a `state` containing mapping information.
+    fn restore(file: &File, state: &GuestMemoryState) -> std::result::Result<Self, Error>;
 }
 
 /// Errors associated with dumping guest memory to file.
 #[derive(Debug)]
 pub enum Error {
+    /// Cannot access file.
+    FileHandle(std::io::Error),
+    /// Cannot create memory.
+    CreateMemory(vm_memory::Error),
+    /// Cannot dump memory.
     WriteMemory(GuestMemoryError),
 }
 
 impl Display for Error {
     fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
         use self::Error::*;
         match self {
-            WriteMemory(err) => write!(f, "Unable to dump memory: {:?}", err),
+            FileHandle(err) => write!(f, "Cannot access file: {:?}", err),
+            CreateMemory(err) => write!(f, "Cannot create memory: {:?}", err),
+            WriteMemory(err) => write!(f, "Cannot dump memory: {:?}", err),
         }
     }
 }
 
-impl DumpMemory for GuestMemoryMmap {
+impl SnapshotMemory for GuestMemoryMmap {
+    /// Describes GuestMemoryMmap through a GuestMemoryState struct.
+    fn describe(&self) -> GuestMemoryState {
+        let mut guest_memory_state = GuestMemoryState {
+            regions: Vec::new(),
+        };
+        let mut offset = 0;
+        let _: std::result::Result<(), ()> = self.with_regions_mut(|_, region| {
+            guest_memory_state.regions.push(GuestMemoryRegionState {
+                base_address: region.start_addr().0,
+                size: region.len() as usize,
+                offset,
+            });
+
+            offset += region.len();
+            Ok(())
+        });
+        guest_memory_state
+    }
+
+    /// Dumps all contents of GuestMemoryMmap to a writer.
     fn dump<T: std::io::Write>(&self, writer: &mut T) -> std::result::Result<(), Error> {
         self.with_regions_mut(|_, region| {
             region
                 .write_to(MemoryRegionAddress(0), writer, region.len() as usize)
                 .map(|_| ())
         })
-        .map_err(Error::WriteMemory)?;
-
-        Ok(())
+        .map_err(Error::WriteMemory)
     }
 
+    /// Dumps all pages of GuestMemoryMmap present in `dirty_bitmap` to a writer.
     fn dump_dirty<T: std::io::Write + std::io::Seek>(
```

Comment:
I think it can be done:
- create a GuestMemoryMmap with a couple of pages and write something on each of them
- create a DirtyBitmap that marks only some of these pages
- check that the resulting file after `dump_dirty` has content (non-zero pages) only for the marked pages

Reply:
sounds good

Reply:
Done.